modules/packages/Sort.chpl

```diff
@@ -18,27 +18,178 @@
+      } else if a < b {
+        return -1;
+      } else {
+        return 0;
+      }
+    }
+
+
+The absolute value comparison example from above can alternatively be
+implemented with a compare method:
+
+.. code-block:: chapel
+
+  var Array = [-1, -4, 2, 3];
+
+  // Empty record serves as comparator
+  record comparator { }
+
+  // compare method defines how 2 elements are compared
+  proc comparator.compare(a, b) {
+    if abs(a) > abs(b) {
+      return 1;
+    } else if abs(a) < abs(b) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  myComparator = new comparator();
+
+  QuickSort(Array, comparator=myComparator);
+
+  // This will output: -1, 2, 3, -4
+  writeln(Array);
+
+If both methods are implemented on the record passed as the comparator, the
+``compare(a, b)`` method will take priority over the ``key(a)`` method.
 
  */
+
 module Sort {
 
-private inline proc chpl_sort_cmp(a, b, param reverse=false, param eq=false) {
-  if eq {
-    if reverse then return a >= b;
-    else return a <= b;
+pragma "no doc"
+/* Empty record to serve as default comparator value in function arguments */
+record defaultComparator{ }
+
+pragma "no doc"
+const emptyRecord = new defaultComparator();
+
+
+pragma "no doc"
+/*
+    Compare function used for all sorts of sort functions.
+
+    Returns:
+      1 if a > b
+      0 if a == b
+     -1 if a < b
+
+ */
+private inline proc chpl_compare(a, b, param reverse=false, comparator:?rec=emptyRecord): int {
+  use Reflection;
+
+  // TODO -- we could also check types
+  // Use comparator.compare(a, b) if is defined by user
+  if canResolveMethod(comparator, "compare", a, b) {
+    if reverse { return -comparator.compare(a, b); }
+    else { return  comparator.compare(a, b); }
+  }
+
+  // TODO -- It may be faster to call key() on all data once, and mirror swaps
+  // Compare results of comparator.key(a) if is defined by user
+  if canResolveMethod(comparator, "key", a) && canResolveMethod(comparator, "key", b) then
+    return chpl_compare(comparator.key(a), comparator.key(b), reverse);
+
+  if reverse {
+    if a < b { return 1; }
+    else if b < a { return -1; }
   } else {
-    if reverse then return a > b;
-    else return a < b;
+    if a < b { return -1; }
+    else if b < a { return 1; }
   }
+
+  return 0;
 }
 
 
 /*
+    Check if a comparator was passed and confirm that it will work, otherwise
+    throw a compile-time error.
+
+    :arg comparator: Comparator record with 'key(a)' or 'compare(a,b)' method
+    :arg a: Sample data passed to confirm that comparator methods can resolve
+    :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+private proc chpl_check_comparator(comparator, a) {
+  use Reflection;
+
+  if comparator.type == defaultComparator {}
+  else if canResolveMethod(comparator, "compare", a, a) {}
+  else if canResolveMethod(comparator, "key", a) {}
```

Comment:
I think it would make sense to do some checks on procs available on the return type of key here as well.